add blog post for the new secret plugin protocol (#2121)
getporter/porter: docs/content/blog/persist-sensitive-data-safely.md

---

> title: "New Secret Plugin Protocol"

Author: The hardest part of it all lol
I will think more on this one

> title: "New Secret Plugin Protocol"

Reviewer: Maybe "Upgrade your plugins to securely store sensitive data"?

> title: "New Secret Plugin Protocol"

Author: I was thinking "Gaining control and security with the new plugin updates", but I like yours better

If you haven't installed the latest Porter v1 release, please follow the instruction [here](https://release-v1.porter.sh/install/#prerelease).

After the installation, let's try to install a bundle that works with sensitive data with just the default configurations in Porter.

```
porter install --reference ghcr.io/getporter/examples/sensitive-data --param password=123a123
```

You should see below error message in the output from the above command:
```
failed to save sensitive param to secrete store: rpc error: code = Unknown desc = The default secrets plugin, secrets.porter.host, does not support persisting secrets: not implemented
```

In this example bundle, there's a sensitive parameter named as `password` and a sensitive output called `name`.

By default, Porter does not provide a built-in solution to persist sensitive data. Any bundle that references or produces sensitive data will fail to execute. 

Now let's configure Porter to persist sensitive data using plain text file through the [filesystem](https://release-v1.porter.sh/plugins/filesystem/) plugin.

```yaml
default-secrets-plugin: "filesystem"
```

> The [filesystem](https://release-v1.porter.sh/plugins/filesystem/) plugin resolves and stores sensitive bundle parameters and outpus as plaintext files in your PORTER_HOME directory.

Reviewer: ```suggestion
The [filesystem plugin](https://release-v1.porter.sh/plugins/filesystem/) resolves and stores sensitive bundle parameters and outpus as plain-text files in your PORTER_HOME directory.
```

Note: the filesystem plugin is only intended for testing and local development usage. It's not intended to be used in production. Please see at the end of this post on how to pick a plugin that is suitable for production usecase.

Now you have a secret store set up, we can try to install the example bundle again.

```
porter install --reference ghcr.io/getporter/examples/sensitive-data --param password=123a123
```

Once the installation process finishes, you should see outputs like below:

```
executing install action from sensitive-data (installation: /sensitive-data)
Install Hello World
Hello, installing example-bundle with password: *******
execution completed successfully!
```

> If you inspect your Porter storage database, in this case, it's a [mongodb instance](https://release-v1.porter.sh/plugins/mongodb-docker/), you should see the value of the `password` parameter is a reference to a [secret](https://release-v1.porter.sh/parameters/#user-specified-values).

Reviewer: The implementation details here are a bit too detailed for a blog post. Let's just leave it at something like "Porter stores a reference to the sensitive data that was saved in the configured secret store."

```
parameters: [
  {
    source: { secret: '01G4DY1RTQVN0DWK67NYJ6TWFH-password' },
    name: 'password'
  }
]
```

> Porter no longer stores the sensitive data through its storage plugins.

Reviewer: Again, I don't think we need to mention the storage plugins specifically, you can just say "in Porter's database"

    Keep sensitive data generated by Porter safe and sound with the new secret plugin protocol
---

As Porter approaches a v1.0.0 release, we have made an improvement in Porter to make sure any sensitive data generated or referenced by Porter is stored in a secure location.
The newly updated secret plugin protocol enables Porter to securely store sensitive data in an external secret store instead of in Porter's database.

> Previously Porter only uses plugins for retrieving secrets from a secret store. When it comes to storing data generated by bundles, Porter uses storage plugins like Mongo as its backend database solution. If a sensitive data, such as a database connection string, were generated by a bundle, it would be stored in a Mongo database in plain text.

Reviewer: ```suggestion
Previously Porter only uses plugins for retrieving secrets from a secret store. When it comes to storing data generated by bundles, Porter uses storage plugins like Mongo as its backend database solution. If sensitive data, such as a database connection string, were generated by a bundle, it would be stored in a Mongo database in plain text.
```

execution completed successfully!
```

> If you inspect your Porter's database, it stores a reference to the sensitive data that was saved in the configured secret store. Porter no longer stores the sensitive data in its database.

Reviewer: ```suggestion
If you inspect Porter's database, it stores a reference to the sensitive data that was saved in the configured secret store. Porter no longer stores the sensitive data in its database.
```
